argh/src/lib.rs

```diff
@@ -455,6 +460,23 @@ impl<T: SubCommand> SubCommands for T {
     const COMMANDS: &'static [&'static CommandInfo] = &[T::COMMAND];
 }
 
+/// Trait implemented by values returned from a dynamic subcommand handler.
+pub trait DynamicCommand<T: Sized>: Send + Sync {
```

Comment:
Would it ever make sense to have a struct DynamicSubcommand? If not, I wonder if it'd make more sense to just focus this on the subcommands interface, and have something like this?

```
pub trait DynamicSubcommands {
    fn commands() -> &'static [&'static CommandInfo] { ... }

    fn redact_arg_values(command_name: &[&str], args: &[&str]) -> Result<Vec<String>, EarlyExit>;

    fn from_args(command_name: &[&str], args: &[&str]) -> Result<Self, EarlyExit>;
}
```

Since we're working with static , I'm not sure if it makes sense to be working with `&self`.

Reply:
The tests create structs that `impl DynamicCommand`. I think you've misunderstood the pattern here, but I'm drawn to your suggestions above and will probably change it anyway :)